locator/topology.hh

```diff
@@ -166,6 +167,82 @@ private:
     friend class topology;
 };
 
+using rack_id = utils::tagged_uuid<struct rack_id_tag>;
+using datacenter_id = utils::tagged_uuid<struct datacenter_id_tag>;
+
+class rack {
+    sstring _name;
+    rack_id _id;
+    std::unordered_set<const node*> _nodes;
+
+    friend class topology;
+    friend class datacenter;
+public:
+    rack() = delete;
+    rack(const rack&) = delete;
+    explicit rack(sstring name) noexcept
+        : _name(std::move(name))
+        , _id(rack_id::create_random_id())
+    {}
+    rack(sstring name, rack_id id) noexcept
+        : _name(std::move(name))
+        , _id(std::move(id))
+    {}
+
+    const sstring& name() const noexcept { return _name; }
+    const rack_id& id() const noexcept { return _id; }
+
+    std::unordered_set<const node*>& nodes() noexcept { return _nodes; }
+    const std::unordered_set<const node*>& nodes() const noexcept { return _nodes; }
+
+    bool empty() const noexcept {
+        return _nodes.empty();
+    }
+    size_t size() const noexcept {
+        return _nodes.size();
+    }
+
+    void for_each_node(std::function<void(const node*)>) const;
+};
+
+class datacenter {
+    sstring _name;
+    datacenter_id _id;
+    std::unordered_map<sstring_view, std::unique_ptr<const rack>> _racks;
+
+    friend class topology;
+public:
+    datacenter() = delete;
+    datacenter(const datacenter&) = delete;
+    explicit datacenter(sstring name) noexcept
+        : _name(std::move(name))
+        , _id(datacenter_id::create_random_id())
+    {}
```

Comment:
Those ids are dangerous, they don't survive restarts and are local to the node.

Reply:
True, but otherwise we have to use names for data structures that can't guarantee stable or cross shard topology instance (holding the datacenter* and rack* objects)

Reply:
We could officially keep them in a system.datacenters table as part of consistent topology and then they could be used cluster wide.
In this case the nodes DC and RACK in system.peers could also store those identifiers.

@tgrabiec, @kbr-scylla WDYT?

Reply:
Let's not add new state, it's difficult enough already. This new state would have two sources (snitch + the tables).

Reply:
What we could do is have a global (not thread-local) grow-only std::unordered_set<datacenter_name> and std::unordered_set<rack_name>. We'd then use pointers/references to these objects.

Reply:
I guess this isn't very different from uuids, except that creating new ones is very explicit.

Reply:
I dropped those identifiers in the latest version.

Reply:
> What we could do is have a global (not thread-local) grow-only std::unordered_set<datacenter_name> and std::unordered_set<rack_name>. We'd then use pointers/references to these objects.

That's doable [1].  Since topology is built first on shard 0 and cloned to other shards we can just point to the objects on shard 0.

If we leak empty data centers and racks pointers to them can be valid forever.
They can be moved aside or be skipped like this series does. Moving aside would be more efficient if empty DCs/RACKs accumulate over time, but if they are repopulated, we should should probably look them up in the "directory" of empty DCs/RACKs and reuse their objects rather than creating new ones.

[1] A standalone set for `std::unordered_set<rack_name>` won't work since rack names aren't globally unique.
It needs to live in the datacenter scope, or be identified by a `tuple<datacenter_name, rack_name>`.